tests/tracer/test_provider.py

```diff
@@ -0,0 +1,218 @@
+import threading
+from typing import Dict
+from typing import Optional
+from typing import Union
+
+from ddtrace.context import Context
+from ddtrace.provider import BaseContextProvider
+from ddtrace.provider import DatadogContextMixin
+from ddtrace.span import Span
+
+
+class TestContextProvider(BaseContextProvider, DatadogContextMixin):
+    """An implementation of a ContextProvider which provides explicit context
+    management through the public attributes `executor_id` and `executor_state`.
+    """
+
+    # The current executor that this provider is in. Change as needed to
+    # simulate other executors.
+    executor_id = 0  # type: int
+
+    # Simulate executor-specific state. This corresponds to thread-locals,
+    # task-locals, etc.
+    executor_state = {}  # type: Dict[int, Optional[Union[Context, Span]]]
+
+    def _has_active_context(self):
+        # type: () -> bool
+        return self.active() is not None
+
+    def active(self):
+        # type: () -> Optional[Union[Context, Span]]
+        ctx = self.executor_state.get(self.executor_id)
+        if isinstance(ctx, Span):
+            return super(TestContextProvider, self)._update_active(ctx)
+        return ctx
+
+    def activate(self, ctx):
+        # type: (Optional[Union[Context, Span]]) -> None
+        self.executor_state[self.executor_id] = ctx
+
+    def _executor_id(self):
+        # type: () -> int
+        return self.executor_id
+
+
+def test_default_cross_execution_parenting_multi_span(tracer, test_spans):
+    """
+    When an active trace is continued in a new executor
+        Spans created in the new executor should inherit from the previous executor.
+    """
+
+    provider = TestContextProvider()
+    tracer.configure(context_provider=provider)
+
+    main_span = tracer.trace("main")
+
+    # Emulate execution going to a new executor (like a thread or task).
+    provider.executor_id = 1
+    # Emulate the context being copied from the previous executor to the new
+    # one.
+    # This should be handled by contextvars (asyncio), monkeypatches (gevent)
+    # or explicit management (threads).
+    provider.activate(main_span.context)
+
+    # New executor begins executing.
+    with tracer.trace("new_executor_1"):
+        pass
+    with tracer.trace("new_executor_2"):
+        pass
+
+    # Execution returns to main executor.
+    main_span.finish()
+
+    main_span = test_spans.find_span(name="main")
+    executor_span1 = test_spans.find_span(name="new_executor_1")
+    executor_span2 = test_spans.find_span(name="new_executor_2")
+
+    assert main_span.trace_id == executor_span1.trace_id
+    assert main_span.trace_id != executor_span2.trace_id
+    assert executor_span1.parent_id == main_span.span_id
+    assert executor_span2.parent_id is None
```

Comment:
Just to clarify, the expected behaviour in a cross-execution tracing scenario is that the first span in the new execution (`executor_span1`) parents from the span in the previous execution (`main_span`)? But the second span created in that new execution (`executor_span2`) does not since it does not directly parent from the first span?

Reply:
"yes", we've talked a bit in the past about whether this should be the case or not. That is, if a span is active and you spawn a new execution, should every span in that new execution be parented to the creator?

Where we end up having problems is with context propagation across processes/the network. e.g. you make an http call from Service A to Service B: Service B handles the request and between finishing the request and handling the next one it does some processing, would that processing also be parented under Service A? Most likely the answer is no, but the implementation ends up being "similar" enough to what we have here that we've decided to hold on any further changes to that behavior (the behavior in this test case is how it works today, so no changes)